File: backend/api/blueprints/workflows.py
"""Mandanten-Workflows (Phase A — Sandbox, pro account_id)."""

from __future__ import annotations

from typing import Any

from flask import Blueprint, g, jsonify, request

from backend.api.middleware.auth_guard import require_auth
from backend.api.middleware.roles import is_account_admin
from backend.api.middleware.tenant import get_request_account_id, require_account
from backend.api.schemas.tenant_workflows import (
    TenantWorkflowCreateRequest,
    TenantWorkflowPreviewRequest,
    TenantWorkflowSuggestRequest,
    TenantWorkflowUpdateRequest,
)
from backend.api.services.tenant_workflow_service import (
    create_workflow,
    delete_workflow,
    get_workflow,
    list_workflows,
    preview_workflow,
    run_workflow_tests,
    suggest_workflow,
    update_workflow,
)

workflows_bp = Blueprint("workflows", __name__, url_prefix="/api/workflows")


def _require_admin() -> tuple[Any, int] | None:
    role = g.current_user.get("role")
    if not is_account_admin(role):
        return jsonify({"error": "Admin required", "code": 403}), 403
    return None


def _account_id() -> str:
    account_id = get_request_account_id()
    assert account_id
    return account_id


@workflows_bp.get("")
@require_auth
@require_account
def tenant_list_workflows() -> tuple[Any, int]:
    data = list_workflows(g.ctx, _account_id())
    return jsonify(data.model_dump()), 200


@workflows_bp.post("")
@require_auth
@require_account
def tenant_create_workflow() -> tuple[Any, int]:
    denied = _require_admin()
    if denied:
        return denied
    body = TenantWorkflowCreateRequest.model_validate(
        request.get_json(silent=True) or {}
    )
    user_id = g.current_user.get("id")
    try:
        created = create_workflow(
            g.ctx,
            _account_id(),
            body,
            user_id=user_id if isinstance(user_id, str) else None,
        )
    except ValueError as exc:
        return jsonify({"error": str(exc)}), 409
    return jsonify(created.model_dump()), 201


@workflows_bp.post("/suggest")
@require_auth
@require_account
def tenant_suggest_workflow() -> tuple[Any, int]:
    denied = _require_admin()
    if denied:
        return denied
    body = TenantWorkflowSuggestRequest.model_validate(
        request.get_json(silent=True) or {}
    )
    suggestion = suggest_workflow(g.ctx, g.settings, body)
    return jsonify(suggestion.model_dump()), 200


@workflows_bp.get("/<workflow_id>")
@require_auth
@require_account
def tenant_get_workflow(workflow_id: str) -> tuple[Any, int]:
    record = get_workflow(g.ctx, _account_id(), workflow_id)
    if record is None:
        return jsonify({"error": "Not found"}), 404
    return jsonify(record.model_dump()), 200


@workflows_bp.put("/<workflow_id>")
@require_auth
@require_account
def tenant_update_workflow(workflow_id: str) -> tuple[Any, int]:
    denied = _require_admin()
    if denied:
        return denied
    body = TenantWorkflowUpdateRequest.model_validate(
        request.get_json(silent=True) or {}
    )
    user_id = g.current_user.get("id")
    try:
        updated = update_workflow(
            g.ctx,
            _account_id(),
            workflow_id,
            body,
            user_id=user_id if isinstance(user_id, str) else None,
        )
    except ValueError as exc:
        return jsonify({"error": str(exc)}), 409
    if updated is None:
        return jsonify({"error": "Not found"}), 404
    return jsonify(updated.model_dump()), 200


@workflows_bp.delete("/<workflow_id>")
@require_auth
@require_account
def tenant_delete_workflow(workflow_id: str) -> tuple[Any, int]:
    denied = _require_admin()
    if denied:
        return denied
    if not delete_workflow(g.ctx, _account_id(), workflow_id):
        return jsonify({"error": "Not found"}), 404
    return jsonify({"deleted": True}), 200


@workflows_bp.post("/<workflow_id>/preview")
@require_auth
@require_account
def tenant_preview_workflow(workflow_id: str) -> tuple[Any, int]:
    denied = _require_admin()
    if denied:
        return denied
    body = TenantWorkflowPreviewRequest.model_validate(
        request.get_json(silent=True) or {}
    )
    result = preview_workflow(g.ctx, g.settings, _account_id(), workflow_id, body)
    if result is None:
        return jsonify({"error": "Not found"}), 404
    return jsonify(result.model_dump()), 200


@workflows_bp.post("/<workflow_id>/run-tests")
@require_auth
@require_account
def tenant_run_workflow_tests(workflow_id: str) -> tuple[Any, int]:
    denied = _require_admin()
    if denied:
        return denied
    result = run_workflow_tests(g.ctx, g.settings, _account_id(), workflow_id)
    if result is None:
        return jsonify({"error": "Not found"}), 404
    return jsonify(result.model_dump()), 200

File: frontend/src/lib/api/workflows.ts
import { apiClient } from "@/lib/api/client";
import type {
  TenantWorkflowCreateRequest,
  TenantWorkflowListResponse,
  TenantWorkflowPreviewRequest,
  TenantWorkflowPreviewResponse,
  TenantWorkflowResponse,
  TenantWorkflowRunTestsResponse,
  TenantWorkflowSuggestRequest,
  TenantWorkflowSuggestResponse,
  TenantWorkflowUpdateRequest,
} from "@/lib/types/api";

export interface WorkflowApiClient {
  fetchWorkflows: () => Promise<TenantWorkflowListResponse>;
  fetchWorkflow: (id: string) => Promise<TenantWorkflowResponse>;
  createWorkflow: (payload: TenantWorkflowCreateRequest) => Promise<TenantWorkflowResponse>;
  updateWorkflow: (
    id: string,
    payload: TenantWorkflowUpdateRequest
  ) => Promise<TenantWorkflowResponse>;
  deleteWorkflow: (id: string) => Promise<void>;
  suggestWorkflow: (
    payload: TenantWorkflowSuggestRequest
  ) => Promise<TenantWorkflowSuggestResponse>;
  previewWorkflow: (
    id: string,
    payload: TenantWorkflowPreviewRequest
  ) => Promise<TenantWorkflowPreviewResponse>;
  runWorkflowTests: (id: string) => Promise<TenantWorkflowRunTestsResponse>;
}

function basePath(adminAccountId?: string): string {
  if (adminAccountId) {
    return `/api/admin/accounts/${adminAccountId}/workflows`;
  }
  return "/api/workflows";
}

export function createWorkflowApi(adminAccountId?: string): WorkflowApiClient {
  const base = basePath(adminAccountId);
  return {
    fetchWorkflows: async () => {
      const { data } = await apiClient.get<TenantWorkflowListResponse>(base);
      return data;
    },
    fetchWorkflow: async (id) => {
      const { data } = await apiClient.get<TenantWorkflowResponse>(`${base}/${id}`);
      return data;
    },
    createWorkflow: async (payload) => {
      const { data } = await apiClient.post<TenantWorkflowResponse>(base, payload);
      return data;
    },
    updateWorkflow: async (id, payload) => {
      const { data } = await apiClient.put<TenantWorkflowResponse>(
        `${base}/${id}`,
        payload
      );
      return data;
    },
    deleteWorkflow: async (id) => {
      await apiClient.delete(`${base}/${id}`);
    },
    suggestWorkflow: async (payload) => {
      const { data } = await apiClient.post<TenantWorkflowSuggestResponse>(
        `${base}/suggest`,
        payload
      );
      return data;
    },
    previewWorkflow: async (id, payload) => {
      const { data } = await apiClient.post<TenantWorkflowPreviewResponse>(
        `${base}/${id}/preview`,
        payload
      );
      return data;
    },
    runWorkflowTests: async (id) => {
      const { data } = await apiClient.post<TenantWorkflowRunTestsResponse>(
        `${base}/${id}/run-tests`
      );
      return data;
    },
  };
}

const tenantApi = createWorkflowApi();

export const fetchWorkflows = tenantApi.fetchWorkflows;
export const fetchWorkflow = tenantApi.fetchWorkflow;
export const createWorkflow = tenantApi.createWorkflow;
export const updateWorkflow = tenantApi.updateWorkflow;
export const deleteWorkflow = tenantApi.deleteWorkflow;
export const suggestWorkflow = tenantApi.suggestWorkflow;
export const previewWorkflow = tenantApi.previewWorkflow;
export const runWorkflowTests = tenantApi.runWorkflowTests;
